### DISCUSSION TITLE
Synchronizing with NTP on a periodic basis

### DISCUSSION THREAD

**Discussion by Ken-Kauffmann** [arnaud-j4k08/PicoClockGreenEasy#4] [Q&A] [answered]
I have read that the temperature-compensated Real Time Clock  can be off by as much as 4 seconds per month. ( The horror ! )

main.cpp contains the follow code snippet before dropping into the main loop:
    ClockUi ui;

    if (Wifi::init())
        if (Wifi::connectBlocking())
            ui.startNtpRequest();

The synchronization is done once, during this startup.  It makes sense to not do it periodically as a default, because you don't want to mess up any user changes to time or date.  

Since the Wifi connection is kept open ( the call to Wifi::deinit() is never called), how difficult would it be to synchronize once per day,  or once per week, say at midnight? Would it be as simple as using the above snippet, but checking for Wifi::status = Connected?  

Where would you suggest placing the code snippet?

Obviously, this is very low priority, as I could just unplug the clock, and plug it back in.

**Comment by arnaud-j4k08** [comment]
> I have read that the temperature-compensated Real Time Clock can be off by as much as 4 seconds per month. ( The horror ! )

Good catch! I have to admit that it may be even worse, as the program only resynchronizes from the RTC at startup. The rest of the time, the time reference is given by the internal timer of the Pico, which may drift even more. However, it's hard to measure, as I reboot my 2 Pico Clock Green's regularly while experimenting new changes.

>Since the Wifi connection is kept open ( the call to Wifi::deinit() is never called), how difficult would it be to synchronize once per day, or once per week, say at midnight? Would it be as simple as using the above snippet, but checking for Wifi::status = Connected?
>
>Where would you suggest placing the code snippet?

A basic version of this feature is easy to implement. In Clock.cpp, at line 105, there is a block executed every minute to handle alarms. Here, you can insert a snippet like this (this one would synchronize at midnight):

```
        if (m_tm.tm_min == 0 && m_tm.tm_hour == 0)
        {
            if (Wifi::linkStatus()== Wifi::Connected)
            {
                startSyncFromNtp();
            }
        }
```

I will consider this enhancement for a future version. A "production ready" solution will require more changes, as such aspects need to be considered:

- To avoid a sudden increase of load on the NTP servers, the synchronization should be done at a random time. Not that I think there will be that many users of PicoClockGreenEasy, but I need to follow the policy for accessing public NTP servers.
- The Wifi connection may have been lost in the meantime, so that a reconnection may be necessary.
- I would like to make the ongoing NTP synchronization visible to the user and I am considering making the "°F" and "°C" indicators blink in this case.
- I would also like to store the time of last synchronization and make it visible to the user somewhere in a menu. Otherwise, I will never be sure if the synchronization has worked.

By the way, I am also working on time synchronization via GPS using an additional GPS module. This will allow clocks without Wifi access to have an accurate time as well.

**Reply by arnaud-j4k08** [reply]
I just confirmed that the current implementation reconnects the Wifi if the connection had been lost in the meantime. The blinking of the "°F" and "°C" indicators is also implemented now. What remains to be done is the daily sync at random time and the "logging" of the last sync.

**Comment by Ken-Kauffmann** [comment]
The use of GPS is a good idea.  I like the wireless connection with NTP, but it does require a user to make changes to the UserConfig.cmake file for SSID and Password.  The GPS could work by having users plug the GPS into pins that you have preconfigured.  That is a nice feature of the PicoGreenClock, to have all the pins available for dupont leads.  

I have experimented with setting the Pico clock with a Goouuu Tech GTU7 from Amazon.  It takes a little bit of time to lock in the first time in a day, but restarts seem to lock in quickly.  My only comment is that you check that you have gotten a good time ( I think that the year initially starts with 2000) before setting clocks.   If I remember correctly, the day of week might be different from 0-->Sunday - 6--> Saturday.

**Reply by arnaud-j4k08** [reply]
I am currently experimenting with a GPS module based on the NEO-6M chip. I connected its RX and TX pins to the GPIO 0 and 1 pins of the Pico, configured as UART. It's indeed nice that the Pico Clock Green gives an easy access to them. The GPS module is small enough to fits in the housing.

Interestingly the Goouuu Tech GTU7 module you mentioned is based on the same NEO-6M chip. Did you write code for it or found something on the internet? I also noticed that the locking time can vary greatly, from 1s to several minutes, depending on when it was done for the last time. I guess it depends on the amount of data that could be reused. The module also has a little battery to maintain the RAM and it can be discharged after a while.

From protocol point of view, the modules sends such a string via UART:
`$GPRMC,135849.00,A,4858.42070,N,00811.11219,E,4.547,353.50,221224,,,A*69`

My existing code extracts the UTC time from "135849 and the date from "221224", whereas the year is only on 2 digits. I did not find any day of week, so it's calculated with mktime in the Clock class.

**Reply by Ken-Kauffmann** [reply]
I haven't written anything from scratch, so I am sure that I was "extending" someone else's work.  Also, it would have been in microPython using Thonny, or using the Arduino IDE with a .ino file.  The PicoClockGreenEasy is my first experience with VisualStudioCode and c/c++.  I think that the imported library was TinyGPS.   It may have provided the Day of Week.

**Comment by arnaud-j4k08** [comment, marked as answer]
RTC, NTP and GPS synchronization on a daily basis is now implemented in version 1.2.0.
